### ISSUE TITLE
ChaCha20/Poly1305 support in Crypto Libraries

### ISSUE THREAD

**Issue by azet** [Homebrew/homebrew#35758]
Hi Guys,

This is just a suggestion I'd like to discuss before opening any kind of PR to see if you're interested in it at all.

I've written to @DomT4 about this on twitter earlier. From talking to very popular CDNs (can't name names unfortunately, but it's easy to guess) as well as Google people; they all compile OpenSSL with the prominent ChaCha20/Poly1305 patch and support it for users accessing customer sites. For NSS (used by Firefox and other software for SSL/TLS) there is also a similar patch. This might be worth thinking about. 

Let me explain in more detail:
One of the reasons Google initially started supporting ChaCha20 (the patch is also written and maintained by them) is that AES-GCM might cause timing problems on mobile devices or devices that do not support the Intel AES-NI instructions. This mainly affects ARM, but might affect other architectures as well. Google seems to have had inhouse research that there is indeed a serious issue. They're not public on details as far as I know, which concerns me a bit. ChaCha20 does not have this issue since it's a constant-time cipher by design. The patch applies to TLS 1.2 and hence adds another AEAD (authenticated encryption with associated data) cipher/mode.

From an IETF perspective: this cipher is not in any standard yet - and certainly will not be added to TLS 1.2 - but is going to be officially supported by TLS 1.3 (I can link to all relevant IETF documents and discussion if really necessary, but this will take me about an hour). 

Support for this is in different Google software, for example Chrome/Chromium:
- https://www.imperialviolet.org/2013/10/07/chacha20.html
- https://code.google.com/p/chromium/issues/detail?id=310768
- https://codereview.chromium.org/59083010/

Android phones make extensive use of this patch for example. Now; modern Macbooks support AES-NI instructions. But as there is a project called `linuxbrew` this might be of interest. It also might be of interest to developers using and linking to OpenSSL/NSS from Homebrew directly. Or Mac users with pre-Westmere Intel processors (i.e. pre-2010 Macbooks).

My suggestion would not be to turn this on by default but make it an option to enable if a user or developer wants support for it. I'd certainly like that for development, for now I custom compile all of these things by hand.

- OpenSSL patch: https://git.openssl.org/gitweb/?p=openssl.git;a=commitdiff;h=9a8646510b
- NSS patch: https://bugzilla.mozilla.org/show_bug.cgi?id=917571

Thanks for your consideration and time,
Aaron

**Comment by DomT4** [comment]
I'll give this a full reply later today.

**Comment by DomT4** [comment]
Okay, so, this is a fun one to answer.

On a personal level, I'd love to merge in this patch ASAP and get ChaCha & Poly support in OpenSSL. I strongly consider them to be some of the most exciting and promising ciphers around at the moment, and I've taken to using them almost everywhere, but there are some not insignificant issues around this.

Firstly, the OpenSSL dev team upstream aren't overly receptive towards AGL's patch right now. In their words, it is "outdated" and "not landing in 1.0.2 immediately". They added that Andy is working on implementing them upstream, but that has not gone into any of the current or immediately future releases yet.

So I'm a little hesitant to roll with something the OpenSSL team considers out-of-date or are going to implement themselves in a different way. Unsupported patches can lead to large offsets, and losing or misconfigured important elements, which in crypto libraries there isn't much room for error, to say the least, heh.

Then there's also the fact that 1.0.2 branch is due for a stable release imminently, and at that point Homebrew will move from 1.0.1 to 1.0.2, hopefully killing SSLv3 in the process, and revisioning things to deal with the API changes and the SSLv3 change. Presumably, the 1.0.2 codebase will be significantly different enough that the patch will break. I'll double check this myself later if there isn't an apparent answer available already.

I'm going to drop AGL an email upstream and consult him on the status of the patch, and whether he'd consider rebasing it against the current codebase. I don't know how much of his time is dedicated to BoringSSL instead these days though. 

If he's willing to draw up an updated patch, and the OpenSSL dev team isn't actively opposed, I have no strong objection to writing it into the formula as strictly an option. However, I don't really want to push a patch in that the OpenSSL dev team considers unready or unacceptable for x, y, z reason, etc, because even supported patches against security software are especially prone to break across versions. *(See for example, the OpenSSH bump in Homebrew/Versions that got delayed significantly whilst a new keychain patch was created)*.

More immediately, I've been trying to migrate some of Homebrew's existing packages to contain LibreSSL as an option, as LibreSSL is a tighter codebase that does support the more modern ciphers. That's only doable where it doesn't require extra patching, as per Homebrew's rules around such things, but we're making small progress on that front.

**Comment by azet** [comment]
HI @DomT4,

Sure - sounds reasonbable.

* ChaCha20/Poly1305 Patch: From what I gather from OpenSSL-dev an IETF lists the real issue for the OpenSSL guys (as well as other implementors) is that chacha20poly1305 is still a draft basically, since TLS 1.3 isn't out there. The support for TLS 1.2 is somewhat of a hack, as I take it it can still be merged because it's quite independent of the rest of the code base (at least that's how ciphers usually are added to OpenSSL, same seems to hold true for this patch - as far as I can see).

* LibreSSL: I strongly recommend against supporting LibreSSL. BoringSSL would be an option since it's upstream OpenSSL with Google's patchset. LibreSSL is known to have security issues on non-OpenBSD plattforms. Also: for example, the eight security vulerabilities released end of last week for OpenSSL and their fixes have not been fixed in LibreSSL to the best of my knowledge. It's a very young project and OpenSSL has gotten steady funding and additional support (3 more full time employees as well as a couple of company volunteers [e.g. people on Googles payroll working on OpenSSL]).

**Comment by DomT4** [comment]
Yeah, I don't mind jumping over the draft status as long as we get a supported patch that upstream aren't dead-set against. If we get a patch that is consistently supported, safe, and doesn't rub upstream the wrong way, I'm happy to write it in as an option, if the maintainers here aren't against that. This largely depends on how AGL and the OpenSSL team feel though - It'd be hard to support the patch against the relatively quickly changing codebase without their support.

On LibreSSL, they had a *lot* of issues at the start, and still have some, but they're making progress as far as I can see. Portability is always complex, and it's a young crypto library and obviously usage of it should be bracketed around its youth. The OpenBSD team have enough security-know-how and funding in place to keep on top of the codebase as it evolves. I think we're being careful enough to explain it's a young library and shouldn't necessarily be relied on in production-vital elements.

LibreSSL *claim* to have fixed the 8 security problems from this week last May. There's a tweet lying around somewhere with the reference, but I can't remember where exactly. BoringSSL is kind of problematic because it doesn't ship with versions yet, it isn't intended as an OpenSSL replacement per AGL's words, and the make process is messy.

The problem rather is, people are fed up with an old, slow, vast codebase in OpenSSL, but expect the same things and more from younger projects whilst maintaining consistency and compatibility. I get the feeling, I understand that, and it's a lot of the reason why you won't find anything in the core that mandates LibreSSL right now. There are a couple of things with LibreSSL choices, but no mandatory dependencies - OpenSSL remains the default, and it will do probably indefinitely.

**Comment by DomT4** [comment]
*(But this probably isn't the place to discuss the various SSL libraries. Let's keep discussion to the merits of the patch, and the possibility of getting it supported across versions).*

**Comment by azet** [comment]
> I'm going to drop AGL an email upstream and consult him on the status of the patch, and whether he'd consider rebasing it against the current codebase. I don't know how much of his time is dedicated to BoringSSL instead these days though.

Excellent.

> (But this probably isn't the place to discuss the various SSL libraries. Let's keep discussion to the merits of the patch, and the possibility of getting it supported upstream & across versions).

I agree. And supporting both projects is actually a nice idea. That being said; apparently they did not fix this issue probably according to OpenSSL developers (see: http://lists.cert.at/pipermail/ach/2015-January/001705.html + two follow up messages). But I digress.

**Comment by denji** [comment]
@azet In LibreSSL ChaCha20/Poly1305 is stable already checked for production. In BoringSSL there are sticky threads at static build (all children process WCPU 100%).

Example nginx + LibreSSL
```
ssl_protocols TLSv1.2 TLSv1.1 TLSv1;
ssl_ciphers ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA:ECDHE-RSA-AES128:DHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA:DES-CBC3-SHA;
ssl_prefer_server_ciphers on;
```

In BoringSSL sampling algorithms is slightly different, here is an example (double containment priority selection)
```
ssl_ciphers [ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305]:[ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305]:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:AES128-GCM-SHA256:AES256-GCM-SHA384:RC4-SHA;
  ssl_ciphers ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA:ECDHE-RSA-AES128:DHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA:DES-CBC3-SHA;
```

**Comment by jacknagel** [comment]
In general we would prefer for open-ended discussion to originate on the mailing list. Once an consensus is reached, then we can work from an issue or pull request.

**Comment by vszakats** [comment]
FYI here's a patch adding ChaCha20-Poly1305 to OpenSSL 1.0.2:
https://github.com/cloudflare/sslconfig/blob/master/patches/openssl__chacha20_poly1305_cf.patch

from this link:
https://blog.cloudflare.com/do-the-chacha-better-mobile-performance-with-cryptography/

**Comment by vszakats** [comment]
There's is pending problems though, patch doesn't apply to 1.0.2 final:
https://github.com/cloudflare/sslconfig/issues/7

**Comment by azet** [comment]
That's because the whole OpenSSL intendation style was changed to kernel style.

See: https://www.openssl.org/blog/blog/2015/02/11/code-reformat-finished/

**Comment by mikemcquaid** [comment]
@vszakats We won't accept a patch until upstream have, FYI.

**Comment by azet** [comment]
@mikemcquaid As far as I understand this is going to end up in upstream anyway, they've genuine interest in that. The patch mentioned above is a bit specific as are other older patches. 

IMHO the best thing here is really to wait it out until a patch is properly reviewed by the OpenSSL team, accepted and merged.

**Comment by vszakats** [comment]
The turbulence it causes is painful, but reformatting was long overdue.
Let's hope the patch clears upstream.

**Comment by mikemcquaid** [comment]
> IMHO the best thing here is really to wait it out until a patch is properly reviewed by the OpenSSL team, accepted and merged to 1.0.2.

:+1:

**Comment by denji** [comment]
LibreSSL supports all these chipers and much better reduce the use of resources and memory.
It can be assigned as a secondary OpenSSL library which is fully compatible with it.

**Comment by azet** [comment]
@denji see my inital comment on LibreSSL: https://github.com/Homebrew/homebrew/issues/35758#issuecomment-69753119

**Comment by vszakats** [comment]
It'd be nice to stay focused on OpenSSL, but these guys have so many eggs in their basket (f.e. maintaining 4-5 major revisions - reduced by two at the end of this year) that even small steps take a long time.

**Comment by azet** [comment]
I'm not betting on LibereSSL at this point. They have far less resources and the OpenBSD guys (as much as I appreciate their work in the security area) have a track record of giving up on some of their projects that were meant as a replacement for something already in existence instead of helping upstream.

**Comment by vszakats** [comment]
@azet I tend to agree, the other major concern is how non-OpenBSD compatibility is considered secondary (if at all). In this regard BoringSSL might be a more viable alternative.

**Comment by denji** [comment]
@vszakats  I've spent the tests BoringSSL on a small number of users in production, unfortunately it sometimes blocks all kernel threads on making SSL connections (BoringSSL it's achievements, but it is still not very stable).

/cc @agl

**Comment by DomT4** [comment]
> how non-OpenBSD compatibility is considered secondary (if at all).

This is largely because making things cross-platform = introducing a load of portability code that's difficult to keep track of and even harder to keep secure. In an ideal world for the BSD devs everyone would use OpenBSD and they'd be able to sit atop an enormously secure hill, I suspect.

**Comment by vszakats** [comment]
@DomT4 Certainly true and I can understand they'd prefer focusing on what they use, know and like most. So LibreSSL can be regarded as an incubator for a clean SSL codebase whilst leaving porting efforts to external contributors. Question: is this happening or will it be happening? (it needs some level of support from OpenBSD guys either way)

**Comment by vszakats** [comment]
[ well, speaking of Homebrew on OS X, the *BSD-ness makes porting a whole lot easier - I guess. ]

**Comment by azet** [comment]
@vszakats Does it? OS X does not ship a BSD kernel, just a BSD userland.

**Comment by vszakats** [comment]
@azet Guessed wrong ;)

**Comment by jacknagel** [comment]
Please take this meta-discussion off the issue tracker, thanks.

**Comment by DomT4** [comment]
> I can understand they'd prefer focusing on what they use, know and like most.

To an extent it's about :heavy_dollar_sign: :heavy_dollar_sign: :heavy_dollar_sign: as well. The OpenBSD team survives on a pretty small amount of money, and portability consequently falls to a lower priority than supporting their own OS.

They [nearly](https://marc.info/?l=openbsd-misc&m=138972987203440&w=2) shut down a lot of their activities last year due to lack of funding, but before last year their Foundation were getting less than ~$35k in donations per year. I guess in this regards they suffer from classic open-source funding issues.

> Question: is this happening or will it be happening?

There's ongoing support for the Portable version of LibreSSL, which is the version Homebrew ships. There have been 6 or 7 releases so far, I believe.

**Comment by DomT4** [comment]
> Please take this meta-discussion off the issue tracker, thanks.

Whoops. This comment did not appear until I hit comment, Sorry :crying_cat_face:

**Comment by denji** [comment]
Google vain threw money on the OpenSSL. LibreSSL opposite corrects all the code with love and takes patches from OpenSSL and BoringSSL.

**Comment by denji** [comment]
There is a proposal to make :libssl and the user chooses openssl/libressl/ etc API/ABI compatible layer.

http://bit.ly/1aLYBqD

**Comment by yqed** [comment]
I released today Nginx 1.9.0 on AXIVO repository, compiled against OpenSSL 1.0.2a with ChaCha+Poly support (RHEL6/7). I'll make later on a detail review, feel free to experiment with the packages. Everything runs live at www.axivo.com.

https://www.axivo.com/packages/setup
https://www.ssllabs.com/ssltest/analyze.html?d=axivo.com&hideResults=on
